tests/dashboards/security/security_dataset_tests.py

```diff
@@ -0,0 +1,260 @@
+
+        db.session.merge(favorite_dash)
+        db.session.merge(regular_dash)
+        db.session.commit()
+
+        dash = db.session.query(Dashboard).filter_by(slug=fav_dash_slug).first()
+
+        favorites = models.FavStar()
+        favorites.obj_id = dash.id
+        favorites.class_name = "Dashboard"
+        favorites.user_id = user.id
+
+        db.session.merge(favorites)
+        db.session.commit()
+
+        self.login(user.username)
+
+        # act
+        get_dashboards_response = self.get_resp(DASHBOARDS_API_URL)
+
+        # assert
+        self.assertIn(f"/superset/dashboard/{fav_dash_slug}/", get_dashboards_response)
+
+    def test_get_dashboards__user_can_not_view_unpublished_dash(self):
+        # arrange
+        admin_user = security_manager.find_user(ADMIN_USERNAME)
+        gamma_user = security_manager.find_user(GAMMA_USERNAME)
+        slug = f"admin_owned_unpublished_dash_{random_slug()}"
+
+        admin_and_not_published_dashboard = create_dashboard_to_db(
+            "My Dashboard", slug=slug, owners=[admin_user]
+        )
+
+        self.login(gamma_user.username)
+
+        # act - list dashboards as a gamma user
+        get_dashboards_response_as_gamma = self.get_resp(DASHBOARDS_API_URL)
+
+        # assert
+        self.assertNotIn(
+            admin_and_not_published_dashboard.url, get_dashboards_response_as_gamma
+        )
+
+    @pytest.mark.usefixtures("load_energy_table_with_slice", "load_dashboard")
+    def test_get_dashboards__users_can_view_permitted_dashboard(self):
+        # arrange
+        username = random_str()
+        new_role = f"role_{random_str()}"
+        self.create_user_with_roles(username, [new_role], copy_roles=True)
+        accessed_table = get_sql_table_by_name("energy_usage")
+        self.grant_role_access_to_table(accessed_table, new_role)
+        # get a slice from the allowed table
+        slice_to_add_to_dashboards = get_slice_by_name("Energy Sankey")
+        # Create a published and hidden dashboard and add them to the database
+        first_dash = create_dashboard_to_db(
+            dashboard_title="Published Dashboard",
+            slug=f"first_dash_{random_slug()}",
+            published=True,
+            slices=[slice_to_add_to_dashboards],
+        )
+
+        second_dash = create_dashboard_to_db(
+            dashboard_title="Hidden Dashboard",
+            slug=f"second_dash_{random_slug()}",
+            published=True,
+            slices=[slice_to_add_to_dashboards],
+        )
+
+        try:
+            self.login(username)
+            # act
+            get_dashboards_response = self.get_resp(DASHBOARDS_API_URL)
+
+            # assert
+            self.assertIn(second_dash.url, get_dashboards_response)
+            self.assertIn(first_dash.url, get_dashboards_response)
+        finally:
+            self.revoke_public_access_to_table(accessed_table)
+
+    @staticmethod
+    def __add_dashboard_mode_params(dashboard_url):
```

Comment:
Could you shed light on where is this function being used and why was it defined in two places?